website/blog/2021-02-04-raptorx.md

```diff
@@ -0,0 +1,152 @@
+---
+title: RaptorX: Building a 10X Faster Presto
+author: James Sun
+authorURL: https://www.linkedin.com/in/yutiansun/
+authorFBID: 100001087056694
+---
+<div style="text-align: justify">
+
+
+**Facebook:** Amit Dutta, Baldeep Hira, Biswapesh Chattopadhyay, James Sun, Jialiang Tan, Ke Wang, Lin Liu, Naveen Cherukuri, Nikhil Collooru, Peter Na, Rohit Jain, Saksham Sachdev, Sergey Pershin, Shixuan Fan, Varun Gajjala
+
+**Alluxio:** Bin Fan, Calvin Jia, Haoyuan Li
+
+**Twitter:** Zhenxiao Luo
+
+**Pinterest:** Lu Niu
+
+*RaptorX is an internal project name aiming to boost query latency significantly faster than vanilla Presto. This blog post introduces the hierarchical cache work, which is the key building block for RaptorX. With the support of the cache, we are able to boost query performance by 10X faster. The architecture can also beat performance-targeted connectors like Raptor.*
+
+
+
+<!--truncate-->
+
+
+## Problems with Disaggregated Storage
+
+[Disaggregate storage](https://en.wikipedia.org/wiki/Disaggregated_storage) is the trend in industry to independently scale storage and compute. It helps cloud providers to save costs. Presto by nature supports such architecture. Data can be streamed from remote storage nodes outside of Presto servers.
+
+
+
+However, storage-compute disaggregation also provides new challenges on query latency as scanning huge amounts of data over the wire is going to be IO bound if we are to saturate the network. Moreover, the metadata paths will also go through the wire to retrieve the location of the data; a few roundtrips of metadata RPCs can easily bump up the latency to more than a second. The following figure shows the IO paths for Hive connectors in orange color, each of which could be a bottleneck on query performance.
+
+
+
+![Presto Architecture](/img/blog/2021-02-04-raptorx/presto-arch.png)
+
+
+
+
+## RaptorX: Build a Hierarchical Caching Solution
+
+Historically, to solve the IO bound issue, Presto has a built-in Raptor connector to load data from remote storage to local SSD for fast access. However, this solution is no different from having a shared compute/storage node, which is against the idea of storage-compute disaggregation. The downside is obvious: either we waste CPU because the SSD in a worker is full or wasting SSD capacity if we are CPU bound. Thus, we started project RaptorX.
+
+
+
+RaptorX is an internal project name aiming to boost query performance for Presto by at least 10X. Hierarchical cache is part of the project. Cache is particularly useful when storage nodes are disaggregated from compute nodes. The goal for RaptorX is not to have a new connector or product, but a built-in solution so that existing workload can seamlessly benefit from it without migration. Especially, we target the existing Hive connector workload.
```

Comment:
```suggestion
RaptorX is an internal project aiming to boost query performance for Presto by at least 10X. Hierarchical caching is the key to RaptorX's success. Cache is particularly useful when storage nodes are disaggregated from compute nodes. The goal of RaptorX is not to create a new connector or product, but a built-in solution so that existing workloads can seamlessly benefit from it without migration. We are specifically targeting the existing Hive connector which is used by many workloads.
```